### PR TITLE
ARROW-14280: [Doc] R package Architectural Overview

### PR DESCRIPTION
**Pull request by AlenkaF**
This PR adds an architectural overview of the R package to the New Contributor's Guide.

This work is co-authored by @thisisnic .

Screenshot 2022-10-03 at 15 09 23

### CODE REVIEW

**Comment by AlenkaF** [comment]
Mind having a look at this PR @jonkeane @thisisnic?
Wasn't sure if this should go to the R dev docs or in the New Contributor's Guide.

**Review by jonkeane — APPROVED** [review]
Thanks for this, this is really great. I'm fine keeping it where it is, but it would be nice to link to it in our developers guide section somewhere too.

The only other thing that I might add is:

* When distributing a source package, we frequently bundle the libarrow source in a directory `tools/cpp` https://github.com/apache/arrow/blob/797f7708e8135ffe7c3a962f4cd27d69e54f3bc8/r/Makefile#L43-L49 for a part of the makefile that does that.
* Do we want to mention anything about the `man` directory with documentation (and that it shouldn't be updated directly?)
* Do we want to mention anything about the vignettes directory?

**Comment by AlenkaF** [comment]
+1 @jonkeane, thank you! Will include suggested and ping when I am done 🙏

**Comment by AlenkaF** [comment]
@jonkeane I think I have addressed all of you comments and the PR should be ready for another round of review.

File: docs/source/developers/guide/architectural_overview.rst [hunk 1]
```diff
@@ -39,5 +39,31 @@ For an Architectural Overview of Arrow's libraries please
...
+         dplyr-funcs*, tools, tests and src/.
+
+* The ``r/R/dplyr-*`` files define the verbs used in a regular
+  dplyr syntax on Arrow objects.
+* The ``r/R/dplyr-funcs*`` files define bindings to Arrow C++
+  functions that can be used with already defined dplyr verbs.
+* All the C++ code connected to the R package lives in ``arrow/r/src``.
+  It also includes C++ code which connects libarrow (the Arrow C++
+  library) and the R code in package.
+* If libarrow source package is bundled with R package using
```
Comment (jonkeane) [inline docs/source/developers/guide/architectural_overview.rst:59]:
```suggestion
* If the libarrow source package is bundled with R package using
```

**Review by jonkeane — APPROVED** [review]
Love this! Thank you so much

**Comment by AlenkaF** [comment]
@thisisnic would be great to have your final review also 🙏

**Review by thisisnic — APPROVED** [review]
LGTM!

**Pull request merged**